finetrainers/trainer.py

```diff
@@ -111,36 +176,189 @@ def prepare_models(self) -> None:
-        logger.info("Initializing dataset and dataloader")
+    def prepare_precomputations(self) -> None:
+        if not self.args.precompute_conditions:
+            return
 
-        self.dataset = VideoDatasetWithResizing(
-            data_root=self.args.data_root,
-            caption_column=self.args.caption_column,
-            video_column=self.args.video_column,
-            resolution_buckets=self.args.video_resolution_buckets,
-            dataset_file=self.args.dataset_file,
-            id_token=self.args.id_token,
+        logger.info("Initializing precomputations")
+
+        if self.args.batch_size != 1:
+            raise ValueError("Precomputation is only supported with batch size 1. This will be supported in future.")
+
+        def collate_fn(batch):
+            latent_conditions = [x["latent_conditions"] for x in batch]
+            other_conditions = [x["other_conditions"] for x in batch]
+            batched_latent_conditions = {}
+            batched_other_conditions = {}
+            for key in list(latent_conditions[0].keys()):
+                if torch.is_tensor(latent_conditions[0][key]):
+                    batched_latent_conditions[key] = torch.cat([x[key] for x in latent_conditions], dim=0)
+                else:
+                    # TODO(aryan): implement batch sampler for precomputed latents
+                    batched_latent_conditions[key] = [x[key] for x in latent_conditions][0]
+            for key in list(other_conditions[0].keys()):
+                if torch.is_tensor(other_conditions[0][key]):
+                    batched_other_conditions[key] = torch.cat([x[key] for x in other_conditions], dim=0)
+                else:
+                    # TODO(aryan): implement batch sampler for precomputed latents
+                    batched_other_conditions[key] = [x[key] for x in other_conditions][0]
+            return {"latent_conditions": batched_latent_conditions, "other_conditions": batched_other_conditions}
+
+        should_recompute = should_perform_precomputation(self.args.data_root)
+        if not should_recompute:
+            logger.info("Precomputed conditions and latents found. Loading precomputed data.")
+            self.dataloader = torch.utils.data.DataLoader(
+                PrecomputedDataset(self.args.data_root),
+                batch_size=self.args.batch_size,
+                shuffle=True,
+                collate_fn=collate_fn,
+                num_workers=self.args.dataloader_num_workers,
+                pin_memory=self.args.pin_memory,
+            )
+            return
+
+        logger.info("Precomputed conditions and latents not found. Running precomputation.")
+
+        # At this point, no models are loaded, so we need to load and precompute conditions and latents
+        condition_components = self.model_config["load_condition_models"](**self._get_load_components_kwargs())
+        self._set_components(condition_components)
+        self._move_components_to_device()
```

Comment:
Because we have to move the modeling components from CPU to GPU, no? It's a handy helper method that moves all currently loaded modules (whether in precomputation stage or training stage) to the GPU